File: backend/app/schemas/sherlock.py
"""Pydantic request/response schemas for Sherlock observability routes.

Phase 15.1d — feeds the platform Logs page's Sherlock tab. ``analytics.log_sherlock_tool_call``
historically had no API surface; the rows were written by the chat handler and
read only by analysts hitting the DB directly. This module exposes a tenant +
user-scoped read API.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Optional

from app.schemas.base import CamelModel, CamelORMModel


class SherlockToolCallRow(CamelORMModel):
    """List-row shape for ``GET /api/sherlock/tool-calls``.

    Heavy payloads (``arguments``, ``generated_sql``, ``validated_sql``) are
    intentionally **not** projected on the list endpoint — the Sherlock tab
    renders a wide table and we don't want to ship MB of JSONB per page.
    The detail endpoint exposes the full row.

    ``args_summary`` is a short comma-joined list of the top-level
    ``arguments`` keys, computed server-side. It gives the table a hint of
    what the tool was called with without serialising the payload.
    """
    id: uuid.UUID
    session_id: Optional[str]
    db_session_id: Optional[uuid.UUID]
    app_id: str
    tool_name: str
    status: str
    error_message: Optional[str]
    execution_ms: Optional[float]
    row_count: Optional[int]
    llm_model: Optional[str]
    cache_hit: Optional[bool]
    args_summary: Optional[str]
    created_at: datetime


class SherlockToolCallDetail(CamelORMModel):
    """Full-row shape for ``GET /api/sherlock/tool-calls/{id}``.

    Includes the raw ``arguments`` payload and both SQL variants. The
    sub-route page renders these in collapsible JSON / SQL sections.
    """
    id: uuid.UUID
    session_id: Optional[str]
    db_session_id: Optional[uuid.UUID]
    app_id: str
    tool_name: str
    status: str
    error_message: Optional[str]
    execution_ms: Optional[float]
    row_count: Optional[int]
    llm_model: Optional[str]
    llm_tokens_in: Optional[int]
    llm_tokens_out: Optional[int]
    cache_hit: Optional[bool]
    arguments: Optional[dict[str, Any]]
    generated_sql: Optional[str]
    validated_sql: Optional[str]
    created_at: datetime


class SherlockToolCallListResponse(CamelModel):
    items: list[SherlockToolCallRow]
    total: int
    limit: int
    offset: int

File: src/services/api/sherlock.ts
/**
 * Phase 15.1d — Sherlock observability client (read-only).
 *
 * Powers the platform Logs page's "Sherlock" tab + the
 * `/<app>/logs/sherlock/:toolCallId` sub-route page. Backend resolves
 * tenant + user scope via the bearer token; the FE only specifies
 * filters and pagination.
 */
import { apiRequest } from './client';

export interface SherlockToolCallRow {
  id: string;
  sessionId: string | null;
  dbSessionId: string | null;
  appId: string;
  toolName: string;
  status: string;
  errorMessage: string | null;
  executionMs: number | null;
  rowCount: number | null;
  llmModel: string | null;
  cacheHit: boolean | null;
  argsSummary: string | null;
  createdAt: string;
}

export interface SherlockToolCallDetail {
  id: string;
  sessionId: string | null;
  dbSessionId: string | null;
  appId: string;
  toolName: string;
  status: string;
  errorMessage: string | null;
  executionMs: number | null;
  rowCount: number | null;
  llmModel: string | null;
  llmTokensIn: number | null;
  llmTokensOut: number | null;
  cacheHit: boolean | null;
  arguments: Record<string, unknown> | null;
  generatedSql: string | null;
  validatedSql: string | null;
  createdAt: string;
}

export interface SherlockToolCallListResponse {
  items: SherlockToolCallRow[];
  total: number;
  limit: number;
  offset: number;
}

export interface ListToolCallsParams {
  appId?: string;
  toolName?: string;
  status?: string;
  sessionId?: string;
  dbSessionId?: string;
  since?: string;
  until?: string;
  limit?: number;
  offset?: number;
}

export async function listToolCalls(
  params?: ListToolCallsParams,
): Promise<SherlockToolCallListResponse> {
  const q = new URLSearchParams();
  if (params?.appId) q.set('appId', params.appId);
  if (params?.toolName) q.set('toolName', params.toolName);
  if (params?.status) q.set('status', params.status);
  if (params?.sessionId) q.set('sessionId', params.sessionId);
  if (params?.dbSessionId) q.set('dbSessionId', params.dbSessionId);
  if (params?.since) q.set('since', params.since);
  if (params?.until) q.set('until', params.until);
  if (params?.limit !== undefined) q.set('limit', String(params.limit));
  if (params?.offset !== undefined) q.set('offset', String(params.offset));
  const qs = q.toString();
  return apiRequest<SherlockToolCallListResponse>(
    `/api/sherlock/tool-calls${qs ? `?${qs}` : ''}`,
  );
}

export async function getToolCall(
  id: string,
  params?: { appId?: string },
): Promise<SherlockToolCallDetail> {
  const q = new URLSearchParams();
  if (params?.appId) q.set('appId', params.appId);
  const qs = q.toString();
  return apiRequest<SherlockToolCallDetail>(
    `/api/sherlock/tool-calls/${id}${qs ? `?${qs}` : ''}`,
  );
}

export async function listDistinctToolNames(params?: { appId?: string }): Promise<string[]> {
  const q = new URLSearchParams();
  if (params?.appId) q.set('appId', params.appId);
  const qs = q.toString();
  return apiRequest<string[]>(
    `/api/sherlock/tool-calls/distinct-tool-names${qs ? `?${qs}` : ''}`,
  );
}
